File: Windows/MediaFlowMonitor/SystemMetrics/ProcessInspector.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;

namespace MediaFlowMonitor.SystemMetrics;

public readonly record struct ProcessUsage(int Pid, string Name, double ValueGB);

/// Top consumatori de RAM/Swap per proces.
///
/// Spre deosebire de macOS (fără câmp public de "swap per proces" — vezi
/// comentariul din ProcessInspector.swift), Windows chiar expune date REALE:
/// `Process.PagedMemorySize64` e octeții din spațiul de adrese al procesului
/// susținuți efectiv de fișierul de paginare (pagefile.sys) — echivalentul
/// direct al coloanei "Commit Size" din Resource Monitor.
public static class ProcessInspector
{
    public static (List<ProcessUsage> Ram, List<ProcessUsage> Swap) TopProcesses(int limit = 3)
    {
        var ram = new List<ProcessUsage>();
        var swap = new List<ProcessUsage>();

        foreach (var process in Process.GetProcesses())
        {
            using (process)
            {
                try
                {
                    double residentGB = process.WorkingSet64 / 1_073_741_824.0;
                    if (residentGB > 0.01)
                        ram.Add(new ProcessUsage(process.Id, process.ProcessName, residentGB));

                    double pagedGB = process.PagedMemorySize64 / 1_073_741_824.0;
                    if (pagedGB > 0.01)
                        swap.Add(new ProcessUsage(process.Id, process.ProcessName, pagedGB));
                }
                catch
                {
                    // Access denied (procese de sistem/alt user) - sarim peste, nu blocam scanarea.
                }
            }
        }

        return (
            ram.OrderByDescending(p => p.ValueGB).Take(limit).ToList(),
            swap.OrderByDescending(p => p.ValueGB).Take(limit).ToList()
        );
    }

    // MARK: - NLE activ (paritate cu ProcessInspector.swift)

    public readonly record struct NLEProcess(string Name, int Pid, double RamGB);

    /// Numele sub care ruleaza efectiv procesele, nu cele din meniul Start.
    /// Potrivire pe fragment, insensibila la majuscule: versiunile difera
    /// („Resolve", „Adobe Premiere Pro 2026"), dar fragmentul ramane.
    private static readonly (string Fragment, string Label)[] NleMarkers =
    {
        ("resolve", "DaVinci Resolve"),
        ("premiere", "Adobe Premiere Pro"),
        ("afterfx", "After Effects"),
        ("mediacomposer", "Avid Media Composer"),
        ("vegas", "VEGAS Pro"),
    };

    /// Aplicatia de montaj activa, daca ruleaza vreuna.
    ///
    /// Se accepta DOAR procesele cu fereastra principala (MainWindowHandle
    /// nenul) — acelasi criteriu ca `activationPolicy == .regular` pe macOS:
    /// un helper de fundal Adobe nu inseamna ca cineva monteaza, iar
    /// badge-ul ar minti.
    public static NLEProcess? ActiveNLE()
    {
        foreach (var process in Process.GetProcesses())
        {
            using (process)
            {
                try
                {
                    string name = process.ProcessName.ToLowerInvariant();
                    var match = NleMarkers.FirstOrDefault(m => name.Contains(m.Fragment, StringComparison.Ordinal));
                    if (match.Label is null) continue;
                    if (process.MainWindowHandle == IntPtr.Zero) continue;
                    return new NLEProcess(match.Label, process.Id, process.WorkingSet64 / 1_073_741_824.0);
                }
                catch
                {
                    // Access denied sau proces iesit intre timp — sarim peste.
                }
            }
        }
        return null;
    }

    // MARK: - DaVinci Resolve zombie detection

    private static List<Process> DavinciProcesses() =>
        Process.GetProcesses().Where(p => p.ProcessName.Contains("Resolve", StringComparison.OrdinalIgnoreCase)).ToList();

    /// True daca exista cel putin un proces "Resolve" cu fereastra vizibila
    /// (MainWindowHandle nenul) - semn ca userul chiar lucreaza in Resolve,
    /// nu ca a mai ramas doar un proces agatat in fundal.
    public static bool IsDaVinciResolveWindowVisible()
    {
        var procs = DavinciProcesses();
        try { return procs.Any(p => { try { return p.MainWindowHandle != IntPtr.Zero; } catch { return false; } }); }
        finally { foreach (var p in procs) p.Dispose(); }
    }

    public static bool AnyDaVinciProcessRunning()
    {
        var procs = DavinciProcesses();
        try { return procs.Count > 0; }
        finally { foreach (var p in procs) p.Dispose(); }
    }

    /// Force Close Hanging DaVinci - kill imediat pe orice proces "Resolve"
    /// gasit. Intoarce numarul de procese omorate.
    public static int ForceKillHangingDaVinci()
    {
        var procs = DavinciProcesses();
        int count = 0;
        foreach (var p in procs)
        {
            using (p)
            {
                try { p.Kill(); count++; }
                catch { /* deja iesit intre timp, sau access denied */ }
            }
        }
        return count;
    }
}

File: macOS/MediaFlowMonitor/Sources/MediaFlowMonitor/SystemMetrics/ProcessInspector.swift
import AppKit
import Darwin
import Foundation

struct ProcessUsage: Identifiable, Equatable {
    let id: Int32 // pid
    let name: String
    let valueGB: Double
}

/// Top consumatori de RAM/Swap per proces.
///
/// IMPORTANT, onest: macOS NU expune public un "swap folosit" per proces —
/// spre deosebire de Windows (PagefileUsage e real). Nu există câmp public
/// în `rusage_info_v4`/`proc_pidinfo` cu octeți efectiv scoși pe disc per
/// proces (nici Activity Monitor nu arată așa ceva). Lista "Swap" de mai
/// jos e deci o aproximare onestă: rangul proceselor după page-in-uri noi
/// (pagini re-aduse din compressor/disk în ultimul interval) — cu cât un
/// proces are mai multe page-ins recente, cu atât SO-ul lucrează mai mult
/// să-i recupereze memoria împinsă afară. Etichetată explicit ca
/// "Swap Activity", nu ca GB static, ca să nu inducă în eroare.
enum ProcessInspector {
    private static var lastPageins: [Int32: UInt64] = [:]
    private static let pageSizeBytes: Double = {
        var size: vm_size_t = 0
        host_page_size(mach_host_self(), &size)
        return Double(size)
    }()

    static func topProcesses(limit: Int = 3) -> (ram: [ProcessUsage], swapActivity: [ProcessUsage]) {
        var pidCount = proc_listallpids(nil, 0)
        guard pidCount > 0 else { return ([], []) }
        var pids = [Int32](repeating: 0, count: Int(pidCount) + 32)
        pidCount = proc_listallpids(&pids, Int32(pids.count) * Int32(MemoryLayout<Int32>.size))

        var ramEntries: [ProcessUsage] = []
        var swapEntries: [ProcessUsage] = []
        var seenThisTick: Set<Int32> = []

        for pid in pids where pid > 0 {
            var info = rusage_info_v4()
            let result = withUnsafeMutablePointer(to: &info) { ptr -> Int32 in
                ptr.withMemoryRebound(to: rusage_info_t?.self, capacity: 1) { rp in
                    proc_pid_rusage(pid, RUSAGE_INFO_V4, rp)
                }
            }
            guard result == 0 else { continue }
            seenThisTick.insert(pid)

            var nameBuf = [CChar](repeating: 0, count: Int(MAXPATHLEN))
            let nameLen = proc_name(pid, &nameBuf, UInt32(nameBuf.count))
            let name = nameLen > 0 ? String(cString: nameBuf) : "pid \(pid)"

            let residentGB = Double(info.ri_phys_footprint) / 1_073_741_824
            if residentGB > 0.01 {
                ramEntries.append(ProcessUsage(id: pid, name: name, valueGB: residentGB))
            }

            let pageins = info.ri_pageins
            let previous = lastPageins[pid] ?? pageins
            let deltaPages = pageins > previous ? pageins - previous : 0
            lastPageins[pid] = pageins
            let deltaMB = Double(deltaPages) * pageSizeBytes / 1_048_576
            if deltaMB > 0.5 {
                swapEntries.append(ProcessUsage(id: pid, name: name, valueGB: deltaMB / 1024))
            }
        }

        // Curăță pid-urile care nu mai există, ca dicționarul să nu crească la infinit.
        lastPageins = lastPageins.filter { seenThisTick.contains($0.key) }

        let topRam = ramEntries.sorted { $0.valueGB > $1.valueGB }.prefix(limit)
        let topSwap = swapEntries.sorted { $0.valueGB > $1.valueGB }.prefix(limit)
        return (Array(topRam), Array(topSwap))
    }

    // MARK: - DaVinci Resolve zombie detection

    /// PID-urile oricărui proces al cărui nume conține "Resolve" — include
    /// procesul principal ȘI helper-ele lui (ex. renderer-e Fusion), care
    /// pot rămâne agățate independent de fereastra principală.
    /// O aplicație de montaj detectată ca rulând acum.
    struct NLEProcess: Equatable {
        let name: String
        let pid: Int32
        let ramGB: Double
    }

    /// Numele sub care rulează efectiv procesele, nu cele din meniul Apple.
    /// Potrivire pe fragment, insensibilă la majuscule: versiunile diferă
    /// („DaVinci Resolve", „Adobe Premiere Pro 2026"), dar fragmentul rămâne.
    private static let nleMarkers: [(fragment: String, label: String)] = [
        ("resolve", "DaVinci Resolve"),
        ("premiere", "Adobe Premiere Pro"),
        ("final cut", "Final Cut Pro"),
        ("after effects", "After Effects"),
        ("avid media composer", "Avid Media Composer"),
    ]

    /// Aplicația de montaj activă, dacă rulează vreuna.
    ///
    /// Se citesc DOAR aplicațiile cu interfață (`NSWorkspace.runningApplications`),
    /// nu tot arborele de procese: un helper de fundal al Adobe nu înseamnă
    /// că cineva montează, iar badge-ul ar minți.
    static func activeNLE() -> NLEProcess? {
        for app in NSWorkspace.shared.runningApplications where app.activationPolicy == .regular {
            guard let name = app.localizedName?.lowercased() else { continue }
            guard let match = nleMarkers.first(where: { name.contains($0.fragment) }) else { continue }
            return NLEProcess(name: match.label, pid: app.processIdentifier, ramGB: residentGB(pid: app.processIdentifier))
        }
        return nil
    }

    /// Memoria rezidentă a unui proces, în GB. 0 dacă nu se poate citi —
    /// badge-ul afișează atunci doar numele și PID-ul, fără consum.
    private static func residentGB(pid: Int32) -> Double {
        var info = proc_taskinfo()
        let size = MemoryLayout<proc_taskinfo>.size
        let result = proc_pidinfo(pid, PROC_PIDTASKINFO, 0, &info, Int32(size))
        guard result == Int32(size) else { return 0 }
        return Double(info.pti_resident_size) / 1_073_741_824
    }

    static func davinciProcessPIDs() -> [Int32] {
        var pidCount = proc_listallpids(nil, 0)
        guard pidCount > 0 else { return [] }
        var pids = [Int32](repeating: 0, count: Int(pidCount) + 32)
        pidCount = proc_listallpids(&pids, Int32(pids.count) * Int32(MemoryLayout<Int32>.size))

        var matches: [Int32] = []
        for pid in pids where pid > 0 {
            var nameBuf = [CChar](repeating: 0, count: Int(MAXPATHLEN))
            let len = proc_name(pid, &nameBuf, UInt32(nameBuf.count))
            guard len > 0 else { continue }
            let name = String(cString: nameBuf)
            if name.localizedCaseInsensitiveContains("resolve") {
                matches.append(pid)
            }
        }
        return matches
    }

    /// True dacă DaVinci Resolve apare ca aplicație "lansată" normal (Dock,
    /// Cmd+Tab) — folosit ca să distingem "userul chiar lucrează în Resolve"
    /// de "procesul a rămas agățat după ce fereastra a fost închisă".
    static func isDaVinciResolveAppVisible() -> Bool {
        NSWorkspace.shared.runningApplications.contains {
            $0.localizedName?.localizedCaseInsensitiveContains("davinci resolve") == true
        }
    }

    /// Închide forțat orice proces "Resolve" agățat. Întoarce numărul de
    /// procese omorâte. SIGKILL — deliberat, e exact acțiunea "Force Close".
    @discardableResult
    static func forceKillHangingDaVinci() -> Int {
        let pids = davinciProcessPIDs()
        for pid in pids { kill(pid, SIGKILL) }
        lastPageins = lastPageins.filter { !pids.contains($0.key) }
        return pids.count
    }
}
